libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/route/FeatureProcessingTask.java

```diff
@@ -14,39 +14,57 @@
 import java.util.ArrayList;
 import java.util.HashMap;
 import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
 
 import static com.mapbox.services.android.navigation.ui.v5.route.RouteConstants.PRIMARY_ROUTE_PROPERTY_KEY;
 
-class FeatureProcessingTask extends AsyncTask<Void, Void, Void> {
+class FeatureProcessingTask extends Thread {
 
   private final List<DirectionsRoute> routes;
   private final List<FeatureCollection> routeFeatureCollections = new ArrayList<>();
   private final WeakReference<OnRouteFeaturesProcessedCallback> callbackWeakReference;
   private final HashMap<LineString, DirectionsRoute> routeLineStrings = new HashMap<>();
+  private AtomicBoolean cancelThread = new AtomicBoolean(false);
+  private Handler postHandler;
 
-  FeatureProcessingTask(List<DirectionsRoute> routes, OnRouteFeaturesProcessedCallback callback) {
+  FeatureProcessingTask(List<DirectionsRoute> routes, OnRouteFeaturesProcessedCallback callback, Handler handler) {
     this.routes = routes;
     this.callbackWeakReference = new WeakReference<>(callback);
+    this.postHandler = handler;
   }
 
   @Override
-  protected Void doInBackground(Void... voids) {
+  public void run() {
     for (int i = 0; i < routes.size(); i++) {
+      if (cancelThread.get()) {
+        return;
+      }
       DirectionsRoute route = routes.get(i);
       boolean isPrimary = i == 0;
       FeatureCollection routeFeatureCollection = createRouteFeatureCollection(route, isPrimary);
       routeFeatureCollections.add(routeFeatureCollection);
     }
-    return null;
+    if (!cancelThread.get()) {
```

Comment:
raw thread has `interrupt` method, which could be called to interrupt the thread, catch exception and do cleanup